src/context.rs

```diff
@@ -0,0 +1,52 @@
+extern crate dotenv;
+extern crate r2d2;
+extern crate r2d2_diesel;
+
+use diesel::sqlite::SqliteConnection;
+use iron::prelude::*;
+use iron::typemap::Key;
+use juniper;
+use persistent::Read;
+use std::error::Error;
+
+pub type ConnectionManager = r2d2_diesel::ConnectionManager<SqliteConnection>;
+pub type Pool = r2d2::Pool<ConnectionManager>;
+pub type PooledConnection = r2d2::PooledConnection<ConnectionManager>;
+
+pub struct IronContext {
+    pub connection_manager: Pool,
+}
+
+#[derive(Copy, Clone)]
+pub struct ContextKey;
+impl Key for ContextKey {
+    type Value = IronContext;
+}
+
+impl IronContext {
+    pub fn init_middleware() -> Result<(Read<ContextKey>, Read<ContextKey>), Box<Error>> {
+        let database_url = dotenv::var("DATABASE_URL")?;
+        let manager = ConnectionManager::new(database_url);
+        let pool: Pool = r2d2::Pool::new(manager)?;
+        let context = IronContext {
+            connection_manager: pool,
+        };
+
+        Ok(Read::<ContextKey>::both(context))
+    }
+}
+
+pub struct GraphQLContext {
+    pub connection: PooledConnection,
```

Comment:
This could be a `SqliteConnection`, so that you don't have to dereference it every time you use it. It might require it to be a reference though, which would be annoying, so I'm not sure how feasible it would be.

Reply:
It needs to be a `PooledConnection` because the fields could potentially be resolved by multiple threads. They probably aren't, but it should be a `PooledConnection` to maintain separation between resolvers.

Reply:
A good way to avoid the dereferencing would be to make the attribute private and add a public method called `connection()` which would do the dereferencing for you.

Reply:
That's basically the same thing except you use a method instead of dereferencing. Not doing it.

Reply:
It is much easier to read, instead of looking at this and wondering what it is really doing:
```rust
let connection = &*context.connection;
```

Reply:
This way, there is no way to get the connection or do anything wrong other than calling the `connection()` method.